cas-client-integration-tomcat-v85/src/main/java/org/jasig/cas/client/tomcat/v85/AbstractAuthenticator.java

```diff
@@ -0,0 +1,200 @@
+ * Version 2.0 (the "License"); you may not use this file
+ * except in compliance with the License.  You may obtain a
+ * copy of the License at the following location:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.jasig.cas.client.tomcat.v85;
+
+import java.io.IOException;
+import java.security.Principal;
+
+import javax.servlet.http.HttpServletResponse;
+
+import org.apache.catalina.*;
+import org.apache.catalina.authenticator.AuthenticatorBase;
+import org.apache.catalina.connector.Request;
+import org.jasig.cas.client.tomcat.AuthenticatorDelegate;
+import org.jasig.cas.client.tomcat.CasRealm;
+import org.jasig.cas.client.util.CommonUtils;
+import org.jasig.cas.client.validation.TicketValidator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Base authenticator for all authentication protocols supported by CAS.
+ * 
+ * @author Scott Battaglia
+ * @version $Revision$ $Date$
+ * @since 3.1.12
+ */
+public abstract class AbstractAuthenticator extends AuthenticatorBase implements LifecycleListener {
+
+    protected final Logger logger = LoggerFactory.getLogger(getClass());
+
+    private final AuthenticatorDelegate delegate = new AuthenticatorDelegate();
+
+    private String casServerUrlPrefix;
+
+    private String encoding;
+
+    private boolean encode;
+
+    private boolean renew;
+
+    protected abstract String getAuthenticationMethod();
+
+    /**
+     * Provided for Tomcat 7.0.8 support.
+     *
+     * @return the authentication method.
+     */
+    protected String getAuthMethod() {
+        return getAuthenticationMethod();
+    }
+
+    /**
+     * Abstract method that subclasses should use to provide the name of the artifact parameter (i.e. ticket)
+     *
+     * @return the artifact parameter name.  CANNOT be NULL.
+     */
+    protected abstract String getArtifactParameterName();
+
+    /**
+     * Abstract method that subclasses should use to provide the name of the service parameter (i.e. service)
+     *
+     * @return the service parameter name.  CANNOT be NULL.
+     */
+    protected abstract String getServiceParameterName();
+
+    /**
+     * Returns the single instance of the ticket validator to use to validate tickets.  Sub classes should include
+     * the one appropriate for the
+     *
+     * @return a fully configured ticket validator.  CANNOT be NULL.
+     */
+    protected abstract TicketValidator getTicketValidator();
+
+    protected void startInternal() throws LifecycleException {
+        super.startInternal();
+        logger.debug("{} starting.", getName());
+        final Realm realm = this.context.getRealm();
```

Comment:
In order to allow access to the realm, needed to add a dependency on tomcat-embedded directly.